e2e/testcases/helm_sync_test.go

```diff
@@ -809,8 +817,174 @@ func TestHelmARTokenAuth(t *testing.T) {
+// 3. Artifact Registry repo: us-docker.pkg.dev/${GCP_PROJECT}/config-sync-test-private
+// 4. GKE cluster with Workload Identity
+// 5. Google Service Account: e2e-test-ar-reader@${GCP_PROJECT}.iam.gserviceaccount.com
+// 6. IAM for the GSA to read from the Artifact Registry repo
+// 7. IAM for the test runner to write to Artifact Registry repo
+// 8. gcloud & helm
+func TestHelmEmptyChart(t *testing.T) {
+	nt := nomostest.New(t,
+		nomostesting.SyncSource,
+		ntopts.Unstructured,
+		ntopts.RequireGKE(t),
+	)
+
+	// Verify workload identity is enabled on the cluster
+	expectedPool := fmt.Sprintf("%s.svc.id.goog", *e2e.GCPProject)
+	if workloadPool, err := getWorkloadPool(nt); err != nil {
+		nt.T.Fatal(err)
+	} else if workloadPool != expectedPool {
+		nt.T.Fatalf("expected workloadPool %s but got %s", expectedPool, workloadPool)
+	}
+
+	// Validate that the GCP Service Account exists.
+	// GSA is used for CS to read the Helm Chart from Artifact Registry.
+	// The local test user will also need to be able to push to Artifact Registry.
+	gsaEmail := gsaARReaderEmail()
+	if _, err := describeGCPServiceAccount(nt, gsaEmail, *e2e.GCPProject); err != nil {
+		nt.T.Fatalf("failed to get service account for workload identity: %v", err)
+	}
+
+	// Create temp dir & delete on exit
+	tempDir, err := os.MkdirTemp("", fmt.Sprintf("%s-", nt.T.Name()))
+	if err != nil {
+		nt.T.Fatalf("failed to create temp directory: %v", err)
+	}
+	nt.T.Cleanup(func() {
+		if err := os.RemoveAll(tempDir); err != nil {
+			nt.T.Errorf("failed to delete temp dir: %v", err)
+		}
+	})
+
+	privateOCIRepo := fmt.Sprintf("us-docker.pkg.dev/%s/config-sync-test-private", *e2e.GCPProject)
+	privateOCIRepoWithScheme := fmt.Sprintf("oci://%s", privateOCIRepo)
+
+	nt.T.Log("Updating Docker config to use gcloud for auth")
+	registry := strings.SplitN(privateOCIRepo, "/", 2)[0]
+	_, err = nt.Shell.ExecWithDebug("gcloud", "auth", "configure-docker", registry)
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+
+	// Use unique chart name:version so it doesn't overlap with parallel runs
+	// of this test in the same GCP project.
+	chartName := strcase.ToKebab(nt.T.Name())
+	chartVersion := time.Now().UTC().Format("20060102.150405.999999999")
+	nt.T.Logf("Creating Helm chart: %s:%s", chartName, chartVersion)
+
+	// Copy chart to temp dir
+	localChartPath := filepath.Join(tempDir, "empty")
+	if err := copyDir(nt, "../testdata/helm-charts/empty", localChartPath); err != nil {
+		nt.T.Fatalf("failed to copy chart: %v", err)
+	}
+	// Modify the chart name and version.
+	localChartFileName := filepath.Join(localChartPath, "Chart.yaml")
+	err = updateYAMLFile(localChartFileName, func(chartMap map[string]interface{}) error {
+		chartMap["name"] = chartName
+		chartMap["version"] = chartVersion
+		return nil
+	})
+	if err != nil {
+		nt.T.Fatalf("failed to update Chart.yaml: %v", err)
+	}
+
+	nt.T.Log("Packaging Helm Chart")
```

Comment:
Do we need to pack and push the empty Helm Chart inside the test?

If the Helm Chart always exists, the test doesn't need the write permission. It also doesn't need to pack and push every time, especially when the test runs on multiple clusters.

Reply:
Ideally the e2e test suite would be runnable in a new project without any manual setup. It's not today, because it assumes a ton of artifacts pre-exist, and we dont have them documented or scripted anywhere.

If we had some system to pre-populate test artifacts, I would use it, but since we don't, I opted to build & push it in the test and also add cleanup and unique versioning to avoid overlap between tests.

Reply:
fwiw, doing this in the test also helps validate that our docs work when explaining how to manage OCI images and Helm charts.

Reply:
Another requirement we might consider in the future: 3rd parties should be able to run all the e2e tests. Today they can't, because they use things like Artifact Registry in our project, which requires setting up IAM to a ServiceAccount that you can use for Workload Identity.

Reply:
We do have scripts and docs for these artifacts: https://github.com/GoogleContainerTools/kpt-config-sync/tree/main/e2e/testinfra#ci-setup.

I believe we need some manual setup anyway, e.g. set-up-kcc-configs.sh and set-up-workload-identity-configs.sh.

For the sake of consistency, I would prefer to create a permanent Helm chart.

It is also not heavy to pack and push the Helm Chart in the test, so I'm okay if you still want to include it in the test.

Reply:
I switched to using RemoteHelmChart to be consistent with the other helm tests.